homeassistant/components/geo_location/__init__.py

```diff
@@ -0,0 +1,153 @@
+        self._hass = hass
+        self._add_devices = add_devices
+        self._name = name
+        self._sort_group_entries_reverse = sort_group_entries_reverse
+        self._managed_devices = []
+        self.group = group.Group.create_group(self._hass, name,
+                                              object_id=util.slugify(name))
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    def _generate_entity_id(self, event_name):
+        """Generate device entity id from the manager name and event name."""
+        entity_ids = [device.entity_id for device in self._managed_devices]
+        entity_id = generate_entity_id(ENTITY_ID_FORMAT,
+                                       '{} {}'.format(self.name, event_name),
+                                       entity_ids, hass=self._hass)
+        return entity_id
+
+    def _group_devices(self):
+        """Re-group all entities."""
+        # Sort entries in group by their state attribute (ascending).
+        devices = sorted(self._managed_devices.copy(),
+                         key=lambda device: device.state,
+                         reverse=self._sort_group_entries_reverse)
+        entity_ids = [device.entity_id for device in devices]
+        # Update group.
+        self.group.update_tracked_entity_ids(entity_ids)
+
+
+class GeoLocationEvent(Entity):
+    """This represents an external event with an associated geo location."""
+
+    def __init__(self, hass, entity_id, name, distance, latitude, longitude,
+                 unit_of_measurement, icon):
+        """Initialize entity with data provided."""
+        self.hass = hass
+        self.entity_id = entity_id
+        self._name = name
+        self._distance = distance
+        self._latitude = latitude
+        self._longitude = longitude
+        self._unit_of_measurement = unit_of_measurement
+        self._icon = icon
+
+    @property
+    def should_poll(self):
+        """No polling needed for a generic geo location event."""
+        return False
+
+    @property
+    def name(self) -> Optional[str]:
+        """Return the name of the event."""
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        """Set event's name."""
+        self._name = value
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return round(self._distance, 1)
+
+    @property
+    def distance(self):
+        """Return distance value of this external event."""
+        return self._distance
+
+    @distance.setter
+    def distance(self, value):
+        """Set event's distance."""
+        self._distance = value
+
+    @property
+    def latitude(self):
+        """Return latitude value of this external event."""
+        return self._latitude
```

Comment:
Return `None`. Platforms should overwrite this.

Reply:
Just wondering what other than the latitude value a platform subclassing this would return instead?

Reply:
Yes. This is the pattern we have for almost all our entity components. We let the platforms define the init of the entity component class and store info in attributes as needed, and the use these in the properties.

Reply:
OK, thanks for the explanation.